Project-based Workflow

Felixmil opened the pull request:
## New Workflow Preview

``` r
devtools::load_all()
#> ℹ Loading esqlabsR
#> Loading required package: ospsuite
#> 
#> Loading required package: rClr
#> 
#> Loading the dynamic library for Microsoft .NET runtime...
#> Loaded Common Language Runtime version 4.0.30319.42000

# Create a project
p <- testProject()

# Project Overview
p
#> 
#> ── Project Configuration ───────────────────────────────────────────────────────
#>   • Project Configuration:
#>   'inst/extdata/examples/TestProject/ProjectConfiguration.xlsx'
#>   • Model folder: 'inst/extdata/examples/TestProject/Models/Simulations'
#>   • Configurations folder: 'inst/extdata/examples/TestProject/Configurations'
#>   • Model Parameters:
#>   'inst/extdata/examples/TestProject/Configurations/ModelParameters.xlsx'
#>   • Individuals:
#>   'inst/extdata/examples/TestProject/Configurations/Individuals.xlsx'
#>   • Populations:
#>   'inst/extdata/examples/TestProject/Configurations/Populations.xlsx'
#>   • PopulationsCSV:
#>   'inst/extdata/examples/TestProject/Configurations/PopulationsCSV'
#>   • Scenarios:
#>   'inst/extdata/examples/TestProject/Configurations/Scenarios.xlsx'
#>   • Applications:
#>   'inst/extdata/examples/TestProject/Configurations/Applications.xlsx'
#>   • Plots: 'inst/extdata/examples/TestProject/Configurations/Plots.xlsx'
#>   • Data folder: 'inst/extdata/examples/TestProject/Data'
#>   • Data file:
#>   'inst/extdata/examples/TestProject/Data/TestProject_TimeValuesData.xlsx'
#>   • Data importer configuration:
#>   'inst/extdata/examples/TestProject/Data/esqlabs_dataImporter_configuration.xml'
#>   • Compound Properties File:
#>   • Output folder: 'inst/extdata/examples/TestProject/Results'
#> 
#> ── Configurations ──────────────────────────────────────────────────────────────
#>   • Scenarios: 5
#>   • Models: 1
#>   • Model Parameters: 3
#>   • Individuals: 1
#> 
#> ── Scenarios ───────────────────────────────────────────────────────────────────
#>   • ✅ TestScenario
#>   • ✅ TestScenario2
#>   • ✅ PopulationScenario
#>   • ✅ PopulationScenarioFromCSV
#>   • ✅ TestScenario_missingParam
#> 
#> ── Simulation Results ──────────────────────────────────────────────────────────
#> No Simulation Results available.

# Explore configurations
p$configurations
#> 
#> ── Configurations ──────────────────────────────────────────────────────────────
#> • Scenarios:
#>   • TestScenario
#>   • TestScenario2
#>   • PopulationScenario
#>   • PopulationScenarioFromCSV
#>   • TestScenario_missingParam
#> • Models:
#>   • Aciclovir.pkml
#> • Model Parameters:
#>   • Global
#>   • MissingParam
#>   • Aciclovir
#> • Individuals:
#>   • Indiv1
#> • Applications:
#>   • Aciclovir_iv_250mg

p$configurations$scenarios$TestScenario
#> $id
#> [1] "TestScenario"
#> 
#> $individual
#> [1] "Indiv1"
#> 
#> $population
#> [1] NA
#> 
#> $populationFromCSV
#> [1] NA
#> 
#> $modelParameters
#> [1] "Global"
#> 
#> $applications
#> [1] "Aciclovir_iv_250mg"
#> 
#> $time
#> [1] "0, 24, 60"
#> 
#> $timeUnit
#> [1] "h"
#> 
#> $steadyState
#> [1] FALSE
#> 
#> $steadyStateTime
#> [1] NA
#> 
#> $steadyStateTimeUnit
#> [1] NA
#> 
#> $model
#> [1] "Aciclovir.pkml"
#> 
#> $outputPaths
#> [1] NA

# Modify configurations
p$configurations$modelParameters$Global$`EHC continuous fraction`$value
#> [1] 1
p$configurations$modelParameters$Global$`EHC continuous fraction`$value <- 0.5
p$configurations$scenarios$TestScenario$modelParameters <- c("Global", "Aciclovir")

# Explore scenarios
p$scenarios$TestScenario$configuration$modelParameters
#> $Global
#> $Global$`EHC continuous fraction`
#> $containerPath
#> [1] "Organism|Liver"
#> 
#> $parameterName
#> [1] "EHC continuous fraction"
#> 
#> $value
#> [1] 0.5
#> 
#> $units
#> [1] ""
#> 
#> 
#> 
#> $Aciclovir
#> $Aciclovir$Lipophilicity
#> $containerPath
#> [1] "Aciclovir"
#> 
#> $parameterName
#> [1] "Lipophilicity"
#> 
#> $value
#> [1] -0.1
#> 
#> $units
#> [1] "Log Units"

# Select which scenarios to run (all active by default)
p$selectScenarios("TestScenario")


# Check the scenarios status
p$scenarios
#> $TestScenario
#> • Status: ✅ active
#> • Type: individual
#> • simulateSteadyTime:
#> • Configurations:
#>   • Model: Aciclovir.pkml
#>   • Model Parameters:
#>     • Global
#>     • Aciclovir
#>   • Individual: Indiv1
#>   • Applications:
#>     • Aciclovir_iv_250mg
#> 
#> $TestScenario2
#> • Status: ⏸ inactive
#> • Type: individual
#> • simulateSteadyTime:
#> • Configurations:
#>   • Model: Aciclovir.pkml
#>   • Model Parameters:
#>     • Global
#>   • Individual: Indiv1
#>   • Applications:
#>     • Aciclovir_iv_250mg
#> 
#> $PopulationScenario
#> • Status: ⏸ inactive
#> • Type: population
#> • simulateSteadyTime:
#> • Configurations:
#>   • Model: Aciclovir.pkml
#>   • Model Parameters:
#>     • Global
#>   • Individual: Indiv1
#>   • Applications:
#>     • Aciclovir_iv_250mg
#> 
#> $PopulationScenarioFromCSV
#> • Status: ⏸ inactive
#> • Type: population
#> • simulateSteadyTime:
#> • Configurations:
#>   • Model: Aciclovir.pkml
#>   • Model Parameters:
#>     • Global
#>   • Individual: Indiv1
#>   • Applications:
#>     • Aciclovir_iv_250mg
#> 
#> $TestScenario_missingParam
#> • Status: ⏸ inactive
#> • Type: individual
#> • simulateSteadyTime:
#> • Configurations:
#>   • Model: Aciclovir.pkml
#>   • Model Parameters:
#>   • Individual: Indiv1
#>   • Applications:
#>     • Aciclovir_iv_250mg

# Run the scenarios
p$runScenarios()
#> Loading Scenarios ■■■■■■■ 20% | ETA: 20s
## This will load and run the scenarios but this can be done in two separated manual steps
# p$loadScenarios()
# modify scenarios
# p$runScenarios()

# Access simulation results
p$simulationResults
#> $TestScenario
#> SimulationResults: 
#>    Number of individuals: 1
```

<sup>Created on 2024-05-06 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

PavelBal committed:
- Fixes #554 (#628)

- Default excel file for population definitions has been renamed from `PopulationParameters.xlsx`
to `Populations.xlsx` (#554, @PavelBal)
- Field `populationParamsFile` of the `ProjectConfiguration` class has been renamed to `populationsFile` (#554, @PavelBal)

PavelBal committed:
Change outputFolder for plots (#630)

* - `createPlotsFromExcel` gets a new argument `outputFolder`. If provided, the plots
are saved in the specified folder. If `NULL`, default results folder as defined
in the `ProjectConfiguration` is used (#495, @PavelBal).

- Function `saveScenarioResults` returns the path where the results were saved.

- Function `createEsqlabsExportConfiguration` requires a new argument `outputFolder`
instead of `projectConfiguration`.

* Fixed vignette

* Do not use `runSimulation`

* Disable partial matching warning for tests

* Getting rid of some test warnings

* Update documnetation of runScenarios

Felixmil committed:
Rename createDefaultProjectConfiguration to createProjectConfiguration without inducing breaking changes
+ usethis::use_package("lifecycle")
+ update tests accordingly
+ add test for deprecation

Felixmil committed:
Harmonize Excel configuration files (#633)

Felixmil committed:
renamed the `Parameters` folder to `Configurations` (#631)

Felixmil committed:
Created helpers for frequently used paths/ProjectConfiguration in tests

Felixmil committed:
fix devtools::document() errors
+ devtools::document()

Felixmil committed:
skip some test on CI (instead of CRAN)

Felixmil committed:
Update NEWS
+ remove names of internal collaborators

Felixmil committed:
Merge remote-tracking branch 'origin/new-project-structure' into paramsFile-to-modelParamsFile

# Conflicts:
#	NEWS.md
#	R/project-configuration.R
#	inst/extdata/examples/TestProject/Configurations/Populations.xlsx
#	inst/extdata/examples/TestProject/ProjectConfiguration.xlsx
#	tests/testthat/test-utilities-population.R

Felixmil committed:
enable GitHub actions when making PR in any branch

Felixmil committed:
remove snapshot for projectConfiguration as it contains absolute path that depends on environment

PavelBal committed:
`NULL` for arguments `simulatedScenarios` and `observedData` in `createPlotsFromExcel` (#629)

* - Function `createPlotsFromExcel` supports `NULL` for arguments `simulatedScenarios`
and `observedData` of no simulation results or observed data are required for
plot generation.

* Typo

* Update docs

Felixmil committed:
start working on project definition

Felixmil committed:
Add populationsCSV path in projectConfiguration
+ use projectConfiguration$populationsCSV to locate population CSV files to read
+ add test to make sure all test scenarios can be properly created

Felixmil committed:
Fix "Indiv" to "Indiv1" to prevent warnings

Felixmil committed:
Merge remote-tracking branch 'origin/new-project-structure' into paramsFile-to-modelParamsFile

# Conflicts:
#	NEWS.md

Felixmil committed:
devtools::document()

Felixmil committed:
Update test expectation

Felixmil committed:
Merge branch 'paramsFile-to-modelParamsFile' into implement-project

Felixmil committed:
Add Project object that contains all esqlabsR workflow components (configurations, scenarios, simulationResults)
+ tests

Felixmil committed:
Improve projectConfiguration print method

Felixmil committed:
style pkg

Felixmil committed:
keep inactivated scenarios status when changing configurations

Felixmil committed:
Improve Project Structure (#635)

* Rename createDefaultProjectConfiguration to createProjectConfiguration without inducing breaking changes
+ usethis::use_package("lifecycle")
+ update tests accordingly
+ add test for deprecation

* Harmonize Excel configuration files (#633)

* renamed the `Parameters` folder to `Configurations` (#631)

* Created helpers for frequently used paths/ProjectConfiguration in tests

* fix devtools::document() errors

* skip some test on CI (instead of CRAN)

* Update NEWS
+ remove names of internal collaborators

* enable GitHub actions when making PR in any branch

* remove snapshot for projectConfiguration as it contains absolute path that depends on environment

* Add populationsCSV path in projectConfiguration
+ use projectConfiguration$populationsCSV to locate population CSV files to read
+ add test to make sure all test scenarios can be properly created

* Fix "Indiv" to "Indiv1" to prevent warnings

* devtools::document()

* Update test expectation

* initialize tlf::ExportConfiguration with passed arguments.
+ link to documentation

* rename populationsCSV to populationsFolder

* Remove compoundPropertiesFile

* Remove compound property in ProjectConfiguration excel file
+ replace leftover previous name

* Correct documentation and object structure
(wasn't doing what it was described to do.)

* Add handler for project configuration from esqlabsR v5
+ general projectConfiguration improvements

* devtools::document()

Felixmil committed:
Correct projectConfiguration related documentation (#653)

Felixmil committed:
ignore excel temporary files

Felixmil committed:
Merge branch 'new-project-structure' into implement-project

# Conflicts:
#	R/project-configuration.R
#	R/scenario.R
#	R/utilities-project-configuration.R
#	inst/extdata/examples/TestProject/ProjectConfiguration.xlsx
#	man/ExportConfiguration.Rd
#	man/ProjectConfiguration.Rd
#	tests/testthat/helpers.R
#	tests/testthat/test-utilities-population.R
#	tests/testthat/test-utilities-project-configuration.R
#	tests/testthat/test-utilities-scenarios.R

Felixmil committed:
Add an optional path argument to save ProjectCOnfiguration
+ add related test

Felixmil committed:
Add test for Project
Small fixes

Felixmil committed:
rename file and relocate functions

Felixmil committed:
Rename files and add "-legacy" to older versions  (temporary)

Felixmil committed:
style + document

Felixmil committed:
replace function with latest name

Felixmil committed:
Better user feedback on simulation results

Felixmil committed:
Add documentation

Felixmil committed:
Add simulationTime objects
+ test
+ documentation

Felixmil committed:
Rename time and timeUnit to simulationTime and simulationTimeUnit
+ add method for nice printing of simulationTime objects

Felixmil committed:
Apply simulationTime output intervals to simulation
+ scenario$configuration is now a copy of configurations$scenarios
+ rename scenarioConfigurationData argument to scenarioConfiguration

Felixmil committed:
Add OutputPaths
+ replace purrr:map(names(... by purrr:imap(...

Felixmil committed:
Add simulationTimeUnit validation

Felixmil committed:
Add steadyState simulation run

Felixmil committed:
Add pretty printing method for individuals

Felixmil committed:
Rename validation function

Felixmil committed:
Rename ModelParameter to SimulationParameter

Felixmil committed:
Add Population object and pop simulation capabilities
+ refactor and document
+ remove legacy code that was re-implemented in new framework
+ relocate "utils" function in appropriate files

Felixmil committed:
devtools::document()

Felixmil committed:
minor fix and optimizations
+devtools::document()

Felixmil committed:
merge main into branch

Felixmil committed:
Merge branch 'main' into new-project-structure

Felixmil committed:
merge branch `main` into `implement-project`

Felixmil committed:
git merge

Merge branch 'new-project-structure' into implement-project

# Conflicts:
#	R/SimulationParameter.R
#	R/utilities-individual.R
#	R/utilities-project-configuration.R
#	R/utilities-scenarios.R
#	R/zzz.R

Felixmil committed:
Merge branch 'new-project-structure' into implement-project

Felixmil committed:
Merge commit 'cf580734044e7a8267bf78a7a6dd67b0667b465d'

#Conflicts:
#	man/ExportConfiguration.Rd
#	tests/testthat/helper-for-tests.R

Felixmil committed:
Merge branch 'new-project-structure' into implement-project

# Conflicts:
#	R/project-configuration.R

Felixmil committed:
Merge branch 'implement-project' of https://github.com/esqLABS/esqlabsR into implement-project

Felixmil committed:
fix save function

Felixmil committed:
Fix tests

Felixmil committed:
Remove duplicated tests

Felixmil committed:
fix bad merge conflict resolution

Felixmil committed:
Restore populationsFolder to populationCSV

Felixmil committed:
remove PI as dependency

Felixmil committed:
Update project directory structure in presentation

Felixmil committed:
Add prerequisite section for installation

Felixmil committed:
Use latest functions in README
+ knit

Felixmil committed:
Rename Population Folder ProjectConfiguration property
- Property is named populationsFolder
- Project template folder contains populations is named PopulationsCSV

Felixmil committed:
styler::style_pkg()

Felixmil committed:
correct object case

Felixmil committed:
Add a test that makes sure that ProjectConfiguration can be customize but throws warning if wront value are provided.

Felixmil committed:
Add NEWS entries
+ format

Felixmil committed:
devtools::document()

Felixmil committed:
extend project configuration tests
+fix typo

AKostiv8 committed:
Update project.R

Felixmil committed:
Prevent user from modifying scenarios directly

PavelBal commented on R/project.R:
```diff
@@ -171,8 +171,9 @@ Project <-
 
         if (!missing(value)) {
-          private$.scenarios <- modifyList(private$.scenarios, value)
+          cli::cli_abort("Scenarios cannot be modified directly. Please configure in project$configurations$scenarios")
```
Message strings must be defined in the messages list.

Felixmil committed:
Refactor repeated error messages

AKostiv8 committed:
config validation

AKostiv8 committed:
Feedback implemented

AKostiv8 committed:
Update project.R

AKostiv8 committed:
Project status tests

AKostiv8 committed:
Test adjustment to the new project structure.

AKostiv8 committed:
update time param warning

Felixmil committed:
fix files names

Felixmil committed:
Improve readme (closes #721)

Felixmil committed:
Update documentation

Felixmil committed:
Remove pksim initialization steps (closes #693)

Felixmil committed:
Merge commit '7d4271027aff77abe19c41cf1749106570fce118'

#Conflicts:
#	NEWS.md

Felixmil committed:
Merge commit '7d4271027aff77abe19c41cf1749106570fce118'

#Conflicts:
#	R/utilities-population.R

Felixmil committed:
replace deprecated function

Felixmil committed:
pull from main

Felixmil committed:
Update path to where results are stored now

Felixmil committed:
Remove embeded shiny apps

Felixmil committed:
Update plots test for new project structure

Felixmil committed:
remove/relocate outdated test

Felixmil committed:
update snaps

AKostiv8 committed:
shiny app remove

AKostiv8 committed:
conflicts resolved

Felixmil committed:
Rename / Comment out of date tests

Felixmil committed:
remove shiny app launching script

Felixmil committed:
merge from main

Felixmil committed:
devtools::documen()

Felixmil committed:
merge new-project-structure into branch

Merge branch 'new-project-structure' into implement-project

# Conflicts:
#	R/utilities-project-configuration.R
#	man/ProjectConfiguration.Rd
#	tests/testthat/helpers.R
#	tests/testthat/test-project-configuration.R
#	tests/testthat/test-utilities-population.R
#	tests/testthat/test-utilities-scenarios.R

AKostiv8 committed:
Merge branch 'individual-validation-demo' into implement-project

AKostiv8 committed:
Merge branch 'implement-project' of https://github.com/esqLABS/esqlabsR into implement-project

AKostiv8 committed:
Revert "Validation object & individual modification validation" (#726)

Felixmil committed:
update snapshot after updating packages

Felixmil committed:
Merge branch 'implement-project' of https://github.com/esqLABS/esqlabsR into implement-project

Felixmil committed:
fix code examples in documentation

Felixmil committed:
fix path to Example Project Configuration File

Felixmil committed:
remove old test files
+ update time unit error regex

Felixmil committed:
fix exampleProject

Felixmil committed:
Move some objects to Internal

Felixmil committed:
disable vignettes (temporary)

Felixmil committed:
restore original build options

Felixmil committed:
remove unsuported character in doc

AKostiv8 committed:
Add Object validation in Project (#728)

* Export Project in a JSON format

As a part of #715

* Import feature added

As a part of #717

* Implement project (#725)

* fix save function

* Fix tests

* Remove duplicated tests

* fix bad merge conflict resolution

* Restore populationsFolder to populationCSV

* remove PI as dependency

* Update project directory structure in presentation

* Add prerequisite section for installation

* Use latest functions in README
+ knit

* Rename Population Folder ProjectConfiguration property
- Property is named populationsFolder
- Project template folder contains populations is named PopulationsCSV

* styler::style_pkg()

* correct object case

* Add a test that makes sure that ProjectConfiguration can be customize but throws warning if wront value are provided.

* Add NEWS entries
+ format

* devtools::document()

* extend project configuration tests
+fix typo

* Prevent user from modifying scenarios directly

* Refactor repeated error messages

* fix files names

* Improve readme (closes #721)

* Update documentation

* Remove pksim initialization steps (closes #693)

* replace deprecated function

* pull from main

* Update path to where results are stored now

* Remove embeded shiny apps

* Update plots test for new project structure

* remove/relocate outdated test

* update snaps

* Rename / Comment out of date tests

* remove shiny app launching script

* merge from main

* devtools::documen()

* Revert "Validation object & individual modification validation" (#726)

* update snapshot after updating packages

---------

Co-authored-by: Felix MIL <felix.mil@esqlabs.com>

* Revert 3b5267764b2d9d86027f217f1a6a69131bbaa7e9

* export project method tests added

* tests update & export modification

---------

Co-authored-by: Felix MIL <felix.mil@esqlabs.com>
Co-authored-by: Felix MIL <34234913+Felixmil@users.noreply.github.com>

AKostiv8 committed:
import plots configuration (#730)

* import plots configuration

* plot configuration object

* Update plot-configuration.R

* replace `dataType` with `label` entry

---------

Co-authored-by: Felix MIL <34234913+Felixmil@users.noreply.github.com>

PavelBal commented:
@Felixmil @AKostiv8 This is the highest prio right now.

Felixmil commented:
I think I can implement the Excel Configurations <-> configurations.json in a separate PR. It will be much faster than finishing this one.
